Extended contribution guide with how to make IDEs recognize Moby’s repository as a Go project

kponichtera opened the pull request:
Closes #43451

@vastjs FYI

**- What I did**

In order to help new contributors configure the project for development in the IDE like GoLand, I have extended the contribution guide with a page, describing what to do to make IDE handle the non-standard structure of the Go project in Moby’s repository (i.e. `vendor.mod` instead of `go.mod`).

The new page can also serve as a good place for the potential future instructions, regarding IDE configuration.

**- How I did it**

I have extended the contribution guide with an optional step, between downloading the required software for development and configuring Git. The reason behind putting it there was that the next step (Git configuration) is where the first file modifications are done, so it would be good to have a functioning IDE by that point.

The added guide page describes the idea behind `vendor.mod` file and `vendor.sh` script. It also mentions that executing the script will generate `go.mod`, necessary for IDE like GoLand to properly recognize the cloned repository as a Go project, index the vendor directory and provide code completion for the libraries.

**- How to verify it**

As @vastjs wrote in https://github.com/moby/moby/issues/43451:

In GoLand 2021.3:
1. Clone repository
2. Open directory in GoLand
3. Open arbitrary Go file (eg. `daemon.go`)
4. Dependencies are not resolved, code completion doesn't work properly
5. Execute `./hack/vendor.sh`
6. GoLand indexes the vendor directory and code completion is now working

**- Description for the changelog**

Extended contribution guide with how to make IDEs recognize Moby’s repository as a Go project

**- A picture of a cute animal (not mandatory but encouraged)**

kponichtera committed:
Extended contribution guide with how to enable GoLand IDE syntax highlighting and code completion in Moby project

Signed-off-by: Konrad Ponichtera <konpon96@gmail.com>

thaJeztah commented on docs/contributing/set-up-ide.md:
```diff
@@ -0,0 +1,33 @@
+will not index dependencies inside the vendor directory.
+As a result, the code completion might be broken.
+
+In order to generate dummy `go.mod` file that will instruct IDE to treat the
+Moby project directory as a Go project, execute the `vendor.sh` script:
```
We should probably not recommend this step; I was actually planning to open a PR to remove the dummy `go.mod` after vendoring is complete, as the `go.mod` causes running tests to fail (due to the `go.mod` being just a "stub").

This project is not (yet) a go module (for various reasons, one of those because of the use of CalVer tags for our binary releases, which go modules doesn't allow).

To configure Goland to work with the code, it's important to:

- check out the source repository in `GOPATH` mode, using `github.com/docker/docker` (which is the name of the packages), e.g. `~/go/src/github.com/docker/docker`


- _disable_ go modules in the Goland configuration for the project:

<img width="1200" alt="Screenshot 2022-04-04 at 19 50 31" src="https://user-images.githubusercontent.com/1804568/161602530-22282862-793d-4dc5-8132-359a6bbb11aa.png">

- And (to prevent the IDE from indexing modules it may find in GOPATH, to disable "Use GOPATH that's defined in system environment" and "index entire GOPATH"

<img width="1200" alt="Screenshot 2022-04-04 at 19 50 31" src="https://user-images.githubusercontent.com/1804568/161603139-67816426-3578-4a5c-8f8e-4942ebf5a28f.png">

kponichtera replied:
Thanks for the tips! I will modify the PR to include these steps instead.

As for removal of the `go.mod` file after vendoring - does it mean that this file is necessary only to run `go mod vendor` command in the `vendor.sh` script? If it is so, would it make sense to remove it in the script's `EXIT` trap, so that it is cleaned up also when some command fails?

thaJeztah replied:
Correct, yes; while `go mod` allows an alternative module-file to be specified (`-modfile=vendor.mod`), it still requires the whole repository to be a module (so a `go.mod` to be present). We had some discussion wether or not the `go.mod` should be removed afterwards and there were _some_ use-cases where it could be useful to keep, I think it's causing more problems than solving them, so I think we should update the script to remove it (for now at least).

We're in the hairy situation currently that (as mentioned) we cannot (yet) move the repository to be a module; adding a `go.mod` to this repository will break all existing users of this repository (`+incompatible` versions are not allowed if the repository has a `go.mod`). So the `vendor.mod` is an "intermediate step" towards being a module, but quite a lot more is needed (we have a path towards that, but details need to be worked out).

kponichtera commented:
@thaJeztah I modified the contribution guide to focus on configuring GoLand instead of creating the dummy `go.mod` file. I have also added a remark for those, who use custom GOPATH directories (for example to have multiple copies of the repository), so that they instruct GoLand explicitly about their location:

![ide_custom_gopath](https://user-images.githubusercontent.com/4958634/165379034-a62f7cf5-d50e-4d9f-a03c-020d9edb4a46.png)

tianon reviewed (approved):
Generally seems OK to me :+1:

thaJeztah reviewed (approved):
LGTM, thanks!

thaJeztah commented:
Jenkins failure is unrelated; let me bring this one in 👍

slonopotamus commented:
I wonder how people worked on Moby before this change at all...

Possibly next step is to add docs on how to run Moby under GoLand's debugger?
